File: 18. KMPStringMatching.py
"""
Task: Implementation and Time Analysis of KMP (Knuth-Morris-Pratt) String Matching Algorithm
"""
def compute_lps(pat):
    m = len(pat)
    lps = [0]*m
    length = 0
    i = 1
    while i < m:
        if pat[i] == pat[length]:
            length += 1
            lps[i] = length
            i += 1
        else:
            if length != 0:
                length = lps[length-1]
            else:
                lps[i] = 0
                i += 1
    return lps
def kmp_search(txt, pat):
    n, m = len(txt), len(pat)
    lps = compute_lps(pat)
    i = j = 0
    while i < n:
        if pat[j] == txt[i]:
            i += 1
            j += 1
        if j == m:
            print(f"Pattern found at index {i-j}")
            j = lps[j-1]
        elif i < n and pat[j] != txt[i]:
            if j != 0:
                j = lps[j-1]
            else:
                i += 1
if __name__ == "__main__":
    import time
    txt = "AABAACAADAABAABA"
    pat = "AABA"
    start = time.time()
    kmp_search(txt, pat)
    end = time.time()
    print(f"Time: {end-start:.6f}s")
"""
Time Complexity: O(n+m)
Space Complexity: O(m)
"""

File: 18. KMPStringMatching.cpp
/*
Task: Implementation and Time Analysis of KMP (Knuth-Morris-Pratt) String Matching Algorithm
*/
#include <iostream>
#include <string>
#include <vector>
#include <ctime>
using namespace std;

void computeLPS(const string& pat, vector<int>& lps)
{
    int m = pat.size(), len = 0;
    lps[0] = 0;
    int i = 1;
    while (i < m)
    {
        if (pat[i] == pat[len])
        {
            lps[i++] = ++len;
        }
        else
        {
            if (len != 0)
                len = lps[len - 1];
            else
                lps[i++] = 0;
        }
    }
}
void KMPSearch(const string& txt, const string& pat)
{
    int n = txt.size(), m = pat.size();
    vector<int> lps(m);
    computeLPS(pat, lps);
    int i = 0, j = 0;
    while (i < n)
    {
        if (pat[j] == txt[i])
        {
            ++i;
            ++j;
        }
        if (j == m)
        {
            cout << "Pattern found at index " << i - j << endl;
            j = lps[j - 1];
        }
        else if (i < n && pat[j] != txt[i])
        {
            if (j != 0)
                j = lps[j - 1];
            else
                ++i;
        }
    }
}
int main()
{
    string txt = "AABAACAADAABAABA";
    string pat = "AABA";
    clock_t start = clock();
    KMPSearch(txt, pat);
    clock_t end = clock();
    cout << "Time: " << (double)(end - start) / CLOCKS_PER_SEC << "s\n";
    return 0;
}
/*
Time Complexity: O(n+m)
Space Complexity: O(m)
*/
